fixes #18681 - moves polymorphic subject into notification object

ohadlevy opened the pull request.

dLobatog reviewed (changes requested):
@ohadlevy Thanks - the reason it got implemented like this is because.. I did not understand what you wanted to do with the subject until we talked about it in person.

Can you add a test that shows the behavior? For example, you may have a 'blueprint.message = "Hello from subject #{notification.subject.name}"'. Then test that by creating a Notification using that NotificationBlueprint with 2 subjects (say, 2 hosts), the message the user gets is:

"Hello from subject host1.example.com"
"Hello from subject host2.example.com"

I'm not sure this is possible to do with the current code in this PR, as the message in the blueprint is static. Perhaps it can be static as "Hello from subject %s", then on the NotificationRecipient you return the message as `_("Hello from subject %s") % notification.subject`.

If that's not the purpose of subject could you explain in a bit more detail what do you want to use it for?

ohadlevy commented:
> If that's not the purpose of subject could you explain in a bit more detail what do you want to use it for?

You are correct, the subject has two use cases.

1. providing a customized messages - e.g. ```  :message: Host %{subject} has been provisioned.```
2. it defines the recipients, for example, we would send the message above only to host owners.

I'll work on implementing 1 as well, 2. is already implemented.

ohadlevy commented:
@dLobatog this is an initial implementation to add a notification that uses subject, I've implemented host built notification.

early feedback is much appreciated.

dLobatog commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,31 @@
+# Namespace for Foreman UI notifications event handling
+module UiNotifications
```
nitpick, why not UINotifications? Lowercase letters ('i' in this case) shouldn't be part of an acronym

dLobatog commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,31 @@
+    attr_reader :state, :subject
+
+    def initialize(subject, state)
+      raise(Foreman::Exception, 'must provide notification subject') if subject.nil?
```
Please translate the string and use `::Foreman::Exception.new(N_("Some error"))` 

Nitpick but the string first letter should be uppercase

dLobatog commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,31 @@
+    def initialize(subject, state)
+      raise(Foreman::Exception, 'must provide notification subject') if subject.nil?
+      @subject = subject
+      raise(Foreman::Exception, 'invalid state') unless allowed_states.include?(state)
```
https://github.com/theforeman/foreman/pull/4339/files#r103540123 , also if you want to validate, how about putting all validations at the beginning? That way we don't set `@subject = subject` for nothing in the case of an invalid `state`

dLobatog commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,31 @@
+    end
+
+    def deliver!
+      raise(Foreman::Exception, 'not Implemented')
```
https://github.com/theforeman/foreman/pull/4339/files#r103540123 , `Not implemented`

dLobatog commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,31 @@
+      User.anonymous_api_admin
+    end
+
+    def allowed_states
```
Please leave a comment explaining why `[]` is here, it's basically forcing the exception in the initializer

dLobatog commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,31 @@
+
+    protected
+
+    def user
```
Just curious, why not `User.anonymous_admin`

dLobatog replied:
Actually.. is this used anywhere? It can probably be removed

dLobatog commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,31 @@
+      []
+    end
+
+    def logger
```
`delegate :logger, :to => :Rails`

dLobatog commented on app/services/ui_notifications/hosts/provision.rb:
```diff
@@ -0,0 +1,51 @@
+      private
+
+      def host_build
+        logger.debug('Host finished provisioning (build is done) - checking for notficiations')
```
typo, `notifications`. Also "build is done" reads a bit weird, "build completed", or "Host built completed (provisioning has finished). Checking for notifications" 

IMO this should be `logger.info`, as it is already for importing facts or reports

Since we have the subject, the message is more traceable by using `Host #{subject}` in the message.

dLobatog commented on app/services/ui_notifications/hosts/provision.rb:
```diff
@@ -0,0 +1,51 @@
+        blueprint = host_build_blueprint
+
+        if blueprint.notifications.where(:subject => subject).empty?
+          logger.debug('creating new UI notification host_build_completed')
```
Ditto, if we have the host name at this point, it'd be better to add it in the message for traceability

dLobatog commented on app/services/ui_notifications/hosts/provision.rb:
```diff
@@ -0,0 +1,51 @@
+            notification_blueprint: blueprint
+          )
+        else
+          logger.debug('UI notification already exists, updating timestamps.')
```
Ditto, if we have the host name at this point, it'd be better to add it in the message for traceability

dLobatog commented on app/services/ui_notifications/hosts/provision.rb:
```diff
@@ -0,0 +1,51 @@
+      def audience
+        if subject.owner_type == 'User'
+          ::Notification::AUDIENCE_USER
+        else
```
If the host has no owner (very common case), there is no notification? Taxonomy could fit in that case

dLobatog commented on app/services/ui_notifications/hosts/provision.rb:
```diff
@@ -0,0 +1,51 @@
+        if blueprint.notifications.where(:subject => subject).empty?
+          logger.debug('creating new UI notification host_build_completed')
+          ::Notification.create!(
+            initiator: User.find(subject.owner_id),
```
`initiator` has to be present, but `subject.owner_id` doesn't... so I don't believe it's a good choice here

dLobatog commented on db/seeds.d/17-notification_blueprints.rb:
```diff
@@ -0,0 +1,30 @@
+  {
+    group: _('Provisioning'),
+    name: 'host_build_completed',
+    message: _('%{subject} has been provisioned successfully'),
```
This is missing `{ :actions => { :links => hosts_path(subject) } ` (probably you'll have to wrap `hosts_path` in a lambda or something though)

dLobatog commented on app/services/ui_notifications/hosts.rb:
```diff
@@ -0,0 +1,6 @@
+module UiNotifications
+  module Hosts
+    class Base < UiNotifications::Base
```
Dead code?

dLobatog reviewed (changes requested):
Thanks @ohadlevy - I tried it and found a few of the issues you can see in the comments, particularly if owner_id is not set. 

To be fair, I think this is a *lot* of YAGNI ceremony for a simple action: sending a notification. I would rather see a:

``` 
class HostBuiltNotificationSender < ApplicationJob
end
```

implementing the `host_build` in `ui_notifications/hosts/provision.rb`. Then in `app/models/host/managed.rb` you just call `HostBuiltNotificationSender.perform_later(self)`, and dynflow picks it up whenever it wants instead of making the action synchronous.

I'm not really opposed to the idea of the UINotifications service but to be honest, I want to keep notifications as simple as possible so people use them, and this (while not complicating notifications on its own) - to a Notifications newbie (most of the people these days) might make them think.. "oh it's not just define a method and call it, it's way more involved"

I don't mind merging if you think it makes sense, but I think in most cases I will resort to ActiveJob and asynchronous notification creation, and keeping stuff contained in its own Job class.

ohadlevy commented:
@dLobatog I believe I've addressed all of your concerns.

I still need to wrap up additional testing the url resolver service.

ohadlevy commented:
/cc @ShimShtein

dLobatog commented on app/services/ui_notifications/url_resolver.rb:
```diff
@@ -0,0 +1,52 @@
+      @raw_actions = actions
+    end
+
+    def message
```
Not sure what this or `parse_string` have to do with URLs?

dLobatog commented on app/services/ui_notifications/hosts/provision.rb:
```diff
@@ -0,0 +1,43 @@
+      end
+
+      def destroyed
+        # IMHO no point in keeping notifications for deleted hosts, as it may
```
nitpick but 'imho' doesn't look like it belongs in a comment

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -1016,4 +1020,8 @@ def send_built_notification
     Foreman::Logging.exception("Host has been created. Failed to send email", e)
   end
+
+  def remove_ui_notifications
```
Is there any way we can have these changes in a concern `Hostext::Notifications` or similar? I'm just worried we keep on adding things to this object. Plus, `Host::Discovered` uses `Hostext` concerns when it needs behavior from `Host::Managed.`

ohadlevy replied:
yes, but can we do it in a separate pr? this one is large enough? :)

dLobatog commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,37 @@
+    end
+
+    def initialize(subject, event)
+      raise(Foreman::Exception, 'must provide notification subject') if subject.nil?
```
Exceptions text should be translated

dLobatog commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,37 @@
+
+    def initialize(subject, event)
+      raise(Foreman::Exception, 'must provide notification subject') if subject.nil?
+      raise(Foreman::Exception, 'invalid event') unless allowed_events.include?(event)
```
Missing translation, also I think it'd be useful to include the event that was passed in the text. `invalid_event %s` or something

dLobatog commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,37 @@
+
+    def deliver!
+      logger.debug("Notification event: #{event} - checking for notifications")
+      send(event) if respond_to?(event, true)
```
Nitpick, `public_send`, or even better, the whole line could just be `try(event)`, correct?

dLobatog commented on app/services/ui_notifications/hosts.rb:
```diff
@@ -0,0 +1,23 @@
+    class Base < UINotifications::Base
+      def deliver!
+        if audience.nil?
+          raise(Foreman::Exception, "Invalid owner for #{subject}, unable to send notifications")
```
Please extract the string for translations

ohadlevy replied:
do you expect to see this anywhere? i can change it to a log entry instead of raising, I was assuming this will be used by ActiveJob so it can halt the job, you are correct that currently with active record call back it might disable the action.

dLobatog commented on app/services/ui_notifications/hosts/provision.rb:
```diff
@@ -0,0 +1,43 @@
+        # IMHO no point in keeping notifications for deleted hosts, as it may
+        # contain links to non existing actions.
+        logger.debug("Removing all notifications for host: #{subject}")
+        Notification.where(subject: subject).delete_all
```
`.destroy_all`, if I'm not wrong `delete_all` is just going to delete stuff in the Notifications table (not NotificationRecipients)

dLobatog commented on app/services/ui_notifications/hosts/provision.rb:
```diff
@@ -0,0 +1,43 @@
+      end
+
+      def initiator
+        return subject.owner if subject.owner.is_a?(User)
```
This is not acknowledging Hosts without owner

dLobatog commented on app/services/ui_notifications/hosts.rb:
```diff
@@ -0,0 +1,23 @@
+          ::Notification::AUDIENCE_USER
+        when 'Usergroup'
+          ::Notification::AUDIENCE_GROUP
+        end
```
```
else
  initiator 
end
```

or something like that - I guess you want the notification even if you don't set the owner field

ohadlevy replied:
no, who would i notify?

ohadlevy replied:
also, the handling is in the deliver! method.

ohadlevy replied:
how about adding a new notification in that case to notify admins they have a host without an owner?

dLobatog reviewed (changes requested):
@ohadlevy Thanks for the changes, please see the comments. 

@domcleal the exceptions here are not to be shown to the user.. we use to translate exceptions everywhere but maybe it doesn't make that much sense in this case?

ShimShtein commented on app/services/ui_notifications/url_resolver.rb:
```diff
@@ -0,0 +1,52 @@
+
+    def parse_link path_method, title
+      {
+        href: public_send(path_method, subject),
```
It will fail for index paths like `hosts_path`, we can't pass the subject to such paths.

``` ruby
hosts_path(Host.first)
#=> "/hosts.tommy-havier.mydomain.com"
```

ShimShtein commented on app/services/ui_notifications.rb:
```diff
@@ -0,0 +1,37 @@
+      User.anonymous_api_admin
+    end
+
+    def allowed_events
```
I would like to see a DSL for extending this list instead of overriding the method. Good example would be `Host` events that could be extended by Katello to add some content specific events.

ohadlevy commented:
@dLobatog would appericiate another review - thanks.

for exceptions, I believe i'll change it away from Foreman::Exception as they should be mostly useful for developers adding notifications vs users?

ShimShtein commented on app/services/ui_notifications/hosts/provision.rb:
```diff
@@ -0,0 +1,43 @@
+      def build_completed
+        blueprint = NotificationBlueprint.find_by(name: 'host_build_completed')
+
+        if blueprint.notifications.where(:subject => subject).empty?
```
You could spare one query:

``` ruby
updated_count = blueprint.notifications.where(subject: subject).update_all(expired_at: blueprint.expired_at)

if updated_count == 0
  ::Notification.create!(...)
else
  log
end
```

ohadlevy commented:
@ShimShtein @dLobatog I think I've addressed all of your concerns?

ShimShtein commented on app/services/ui_notifications/hosts/build_completed.rb:
```diff
@@ -0,0 +1,36 @@
+      private
+
+      def create
+        if blueprint.notifications.where(:subject => subject).empty?
```
Nitpick, but we can spare this query:

``` ruby
if update_notifications() == 0 #no old notifications
  add_notification
end
```

ShimShtein commented on app/services/ui_notifications/hosts/build_completed.rb:
```diff
@@ -0,0 +1,36 @@
+      end
+
+      def blueprint
+        @blueprint ||= NotificationBlueprint.find_by(name: 'host_build_completed')
```
We can cache this blueprint on class level - it will spare a query for each notification.

ohadlevy replied:
@ShimShtein but the blueprint is per class? each class will have its own?

ShimShtein commented on app/services/ui_notifications/hosts/destroy.rb:
```diff
@@ -0,0 +1,29 @@
+      end
+
+      def blueprint
+        @blueprint ||= NotificationBlueprint.find_by(name: 'host_destroyed')
```
ditto

ShimShtein commented on app/services/ui_notifications/url_resolver.rb:
```diff
@@ -0,0 +1,61 @@
+  class URLResolver
+    # needed in order to resolve rails urls for notifications
+    include Rails.application.routes.url_helpers
+    def initialize subject, actions = nil
```
parenthesis PLZ

ohadlevy replied:
updated, even if rubocop does not enforce this.

ohadlevy commented:
@ShimShtein updated based on your comments.

ohadlevy commented:
adding a screenshot for the 3 new notifications:
1. host finished provisioning
2. host has no owner set
3. host is destroyed
![notifications](https://cloud.githubusercontent.com/assets/20167/23697895/308b07d6-03f2-11e7-91b9-ac87b71f1e93.png)

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -50,7 +51,7 @@ def initialize(*args)
 
   def build_hooks
-    return unless respond_to?(:old) && old && (build? != old.build?)
+    return if previous_changes['build'].nil?
```
nitpick, `blank`

ohadlevy replied:
this can be nil or an array, empty string / array is not possible for this method.

dLobatog commented on app/models/notification.rb:
```diff
@@ -61,4 +65,17 @@ def set_notification_recipients
+      notification_blueprint.actions
+    ).actions if notification_blueprint.actions.any?
+    # copy notification message in case we didn't create a custom one.
+    self.message ||= UINotifications::StringParser.new(
```
Nice separation of concerns

dLobatog commented on app/services/ui_notifications/hosts/destroy.rb:
```diff
@@ -0,0 +1,29 @@
+
+      def create
+        # I'm defaulting to deleting older notifications as it may
+        # contain links to non existing actions.
```
👍

dLobatog commented on app/services/ui_notifications/hosts/missing_owner.rb:
```diff
@@ -0,0 +1,29 @@
+      def add_notification
+        Notification.create!(
+          initiator: initiator,
+          audience: 'admin',
```
nitpick, `Notification::AUDIENCE_ADMIN` to be resilient to changes and use the frozen string

dLobatog reviewed (changes requested):
@ohadlevy Code-wise 👍 , I'm going to test this a bit more thoroughly and merge if it works. Thanks!

ohadlevy commented:
@dLobatog or others, do you have any idea why the tests are failing? I'm unsure why the hook is not firing / updating correctly.

ohadlevy committed:
fixes #18681 - moves polymorphic subject into notification object

also adds host build, destroyed and missing owner UI notifications

ohadlevy commented:
@dLobatog I think I've addressed all failing tests. I've also added exception handling in the case notification action fails, I do not want to block the caller (e.g. think AR callbacks). once we introduce ActiveJob this might change, but regardless I think its good idea not to block the UI/Action triggering the notification regardless.

ohadlevy commented:
@dLobatog to make sure, the lines I've added are at https://github.com/theforeman/foreman/pull/4339/files#diff-59ce2b15fb9e1d3ecb10bf58cf982db8R8

dLobatog commented:
Thanks @ohadlevy ! Feel free to start using this in discovery, I'll do the same.
